tests/network/upgrade/test_localnet_connectivity.py

```
@@ -0,0 +1,97 @@
+        - TCP connection succeeds.
+    """
+    vm_a, vm_b = localnet_running_vms
+    iface = lookup_iface_status(vm=vm_b, iface_name=LOCALNET_BR_EX_INTERFACE)
+    for dst_ip in filter_link_local_addresses(ip_addresses=iface.ipAddresses):
+        with subtests.test(msg=f"IPv{dst_ip.version}"):
+            with client_server_active_connection(
+                client_vm=vm_a,
+                server_vm=vm_b,
+                spec_logical_network=LOCALNET_BR_EX_INTERFACE,
+                ip_family=dst_ip.version,
+            ) as (client, server):
```

Ruff (0.15.15)

[warning] Use a single `with` statement with multiple contexts instead of nested `with` statements

(SIM117)